fedimint-core/src/db/mod.rs

```diff
@@ -271,7 +271,34 @@ pub trait IDatabaseTransaction<'a>: 'a + Send {
         Ok(())
     }
 
-    async fn commit_tx(self: Box<Self>) -> Result<()>;
+    async fn commit_tx(self) -> Result<()>;
+
+    async fn rollback_tx_to_savepoint(&mut self);
+
+    /// Create a savepoint during the transaction that can be rolled back to
+    /// using rollback_tx_to_savepoint. Rolling back to the savepoint will
+    /// atomically remove the writes that were applied since the savepoint
+    /// was created.
+    ///
+    /// Warning: Avoid using this in fedimint client code as not all database
+    /// transaction implementations will support setting a savepoint during
+    /// a transaction.
+    async fn set_tx_savepoint(&mut self);
+}
+
+#[async_trait]
+pub trait IDatabaseTransaction<'a>: 'a + Send {
```

Comment:
I though we're going to `trait IDatabaseTransaction: IDatabaseTransactionNoCommit` or something like that.

Reply:
One way or another `Core` doesn't tell much.

Reply:
Ah, this not about lack of commit, but `&mut self` vs `self`.

Reply:
Yes this only handles `&mut self` vs `self`.

We can still split the trait up into `IDatabaseTransaction` and `IDatabaseTransactionCommit` if we want, but I worry introducing more traits will just make it more confusing. Perhaps the better solution is just to introduce the new public wrapper struct that can't commit and give that to the modules. I think this pretty much accomplishes the same goal, but I'm open to suggestions.